crates/outbound-mqtt/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+mod host_component;
+
+use std::time::Duration;
+
+use anyhow::Result;
+use paho_mqtt::Client;
+use spin_core::{async_trait, wasmtime::component::Resource};
+use spin_world::v1::mqtt as v1;
+use spin_world::v2::mqtt::{self as v2, Connection as MqttConnection, Error, Qos};
+
+pub use host_component::OutboundMqttComponent;
+
+pub struct OutboundMqtt {
+    allowed_hosts: spin_outbound_networking::AllowedHostsConfig,
+    connections: table::Table<Client>,
+}
+
+impl Default for OutboundMqtt {
+    fn default() -> Self {
+        Self {
+            allowed_hosts: Default::default(),
+            connections: table::Table::new(1024),
+        }
+    }
+}
+
+impl OutboundMqtt {
+    fn is_address_allowed(&self, address: &str) -> bool {
+        spin_outbound_networking::check_url(address, "mqtt", &self.allowed_hosts)
+    }
+
+    async fn establish_connection(
+        &mut self,
+        address: String,
+        username: String,
+        password: String,
+        keep_alive_interval: Duration,
+    ) -> Result<Result<Resource<MqttConnection>, Error>> {
+        Ok(async {
+            let client = Client::new(address.as_str()).map_err(|_| Error::InvalidAddress)?;
+            let conn_opts = paho_mqtt::ConnectOptionsBuilder::new()
+                .keep_alive_interval(keep_alive_interval)
+                .user_name(username)
+                .password(password)
+                .finalize();
+
+            client.connect(conn_opts).map_err(other_error)?;
+
+            self.connections
+                .push(client)
+                .map(Resource::new_own)
+                .map_err(|_| Error::TooManyConnections)
+        }
+        .await)
+    }
+}
+
+impl v2::Host for OutboundMqtt {}
+
+#[async_trait]
+impl v2::HostConnection for OutboundMqtt {
+    async fn open(
+        &mut self,
+        address: String,
+        username: String,
+        password: String,
+        keep_alive_interval: u64,
+    ) -> Result<Result<Resource<MqttConnection>, Error>> {
+        if !self.is_address_allowed(&address) {
+            return Ok(Err(Error::InvalidAddress));
```

Comment:
This should specifically indicate that the address is not permitted (cf. e.g https://github.com/fermyon/spin/blob/3ee465e8ae114d627a1441327bca44d5c9ad0084/crates/outbound-mysql/src/lib.rs#L84)

Reply:
Agreed, more appropriate (will update), think we may need to reflect this in [outbound Redis](https://github.com/fermyon/spin/blob/3ee465e8ae114d627a1441327bca44d5c9ad0084/crates/outbound-redis/src/lib.rs#L76) too for consistency?

Reply:
That would be good, but I don't think Redis has a suitable error variant for doing that, and adding or modifying one would have compatibilty considerations.  Longer term for sure though - thanks for the spot.